bpo-33530: Implement Happy Eyeballs in asyncio.

twisteroidambassador opened the pull request:
Added two keyword arguments, `delay` and `interleave`, to
`BaseEventLoop.create_connection`. Happy eyeballs is activated if
`delay` is specified.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: bpo-33530 -->
https://bugs.python.org/issue33530
<!-- /issue-number -->

twisteroidambassador committed:
Implement Happy Eyeballs in asyncio.

Added two keyword arguments, `delay` and `interleave`, to
`BaseEventLoop.create_connection`. Happy eyeballs is activated if
`delay` is specified.

twisteroidambassador committed:
Implement Happy Eyeballs in asyncio.

Added two keyword arguments, `delay` and `interleave`, to
`BaseEventLoop.create_connection`. Happy eyeballs is activated if
`delay` is specified.

1st1 commented on Lib/asyncio/helpers.py:
```diff
@@ -0,0 +1,142 @@
+from contextlib import suppress
```
I'm usually against using module names like "helpers" or "utils", as it's impossible to know what's in them by just looking at the name. I propose to call this module "staggered.py".

1st1 replied:
Please also import modules, not functions/classes directly. That's the style for asyncio code we try to maintain.

1st1 commented on Lib/asyncio/base_events.py:
```diff
@@ -839,12 +874,49 @@ def _check_sendfile_params(self, sock, file, offset, count):
             local_addr=None, server_hostname=None,
-            ssl_handshake_timeout=None):
+            ssl_handshake_timeout=None,
+            delay=None, interleave=None):
```
Need to mention these in the docstring.  

~Is it a good idea to have one parameter instead of two?  For example, `interleave=-1` (default) is used to disable happy eyeballs. `interleave>0` is used as a delay for happy eyeballs.~

Oh, I misunderstood the meaning of `interleave`. Why shouldn't we always interleave addrs?

twisteroidambassador replied:
> Why shouldn't we always interleave addrs?

Mostly because RFC8305 specifies "First Address Family Count" as a configurable value, and since I'm making it configurable, might as well provide the option of turning it off. Also, address family interleaving is a "SHOULD" in the RFC. If `interleave` is not specified, the default behavior is to interleave.

If we decide that the use case for configuring `interleave` is too small, we could remove it outright and just always interleave.

twisteroidambassador replied:
> Need to mention these in the docstring.

The docstring of `create_connection` itself? The existing docstring doesn't explain all those old arguments, so I wondered whether it's appropriate to mention the new one.

(The larger Python docs should probably be updated as well?)

vstinner committed:
bpo-33629: Prevent coredump in test_importlib (GH-7090)

bpo-32374, bpo-33629: Use support.SuppressCrashReport() in
test_bad_traverse() of MultiPhaseExtensionModuleTests to prevent
leaking a core dump file.

serhiy-storchaka committed:
bpo-32493: Fix uuid.uuid1() on FreeBSD. (GH-7099)

Use uuid_enc_be() if available to encode UUID to bytes as big endian.

ilevkivskyi committed:
Document typing.NoReturn (GH-7107)

vstinner committed:
test.regrtest: flush stdout when display progress (#7105)

runtest_mp.py: call print() with flush=True.

serhiy-storchaka committed:
bpo-33644: Fix signatures of tp_finalize handlers in testing code. (GH-7111)

andresdelfino committed:
bpo-32769: A new take on annotations/type hinting glossary entries (GH-6829)

andresdelfino committed:
Fix typo spotted by Guido (GH-7131)

serhiy-storchaka committed:
bpo-33652: Improve pickle support in the typing module. (GH-7123)

Pickles of type variables and subscripted generics are now future-proof
and compatible with older Python versions.

ned-deily committed:
bpo-33655: Also ignore test_posix_fallocate failures on BSD platforms (GH-7134)

The failure may be due to the use oF ZFS, a case we already ignore
for Solaris-based systems where ZFS is frequently used.

emmatyping committed:
bpo-32380: Create functools.singledispatchmethod (#6306)

serhiy-storchaka committed:
bpo-33652: Remove __getstate__ and __setstate__ methods in typing. (GH-7144)

Commit:
bpo-32374: Ignore Python-level exceptions in test_bad_traverse (GH-7145)

Commit:
bpo-33623: Fix possible SIGSGV when asyncio.Future is created in __del__ (#7080)

twisteroidambassador committed:
bpo-31647: Fix write_eof() after close() for SelectorSocketTransport (GH-7149)

Fixed bug where calling write_eof() on a _SelectorSocketTransport after
it's already closed raises AttributeError.

Commit:
bpo-33505: Optimize asyncio.ensure_future by reordering if conditions (GH-6836)

Commit:
bpo-33654: Support BufferedProtocol in set_protocol() and start_tls() (GH-7130)

In this commit:

* Support BufferedProtocol in set_protocol() and start_tls()
* Fix proactor to cancel readers reliably
* Update tests to be compatible with OpenSSL 1.1.1
* Clarify BufferedProtocol docs
* Bump TLS tests timeouts to 60 seconds; eliminate possible race from start_serving
* Rewrite test_start_tls_server_1

vstinner committed:
regrtest: repeat summary after re-run (GH-7159)

Using -w, when failing tests are re-run in verbose mode, display
again the tests results at the end.

zooba committed:
bpo-33614: Ensures module definition files for the stable ABI on Windows are correctly regenerated. (GH-7165)

Commit:
bpo-32458: Further tune start_tls tests (#7166)

Commit:
bpo-33672: Fix Task.__repr__ crash with Cython's bogus coroutines (GH-7161)

zooba committed:
bro-33614: Override exit code in find_msbuild.bat (GH-7169)

Commit:
bpo-33469: RuntimeError after closing loop that used run_in_executor (GH-7171)

Commit:
bpo-32610: Fix asyncio.all_tasks() to return only pending tasks. (GH-7174)

Commit:
bpo-32410: Avoid blocking on file IO in sendfile fallback code (GH-7172)

andresdelfino committed:
bpo-33673: Install python-docs-theme even if Sphinx is already installed (GH-7163)

andresdelfino committed:
bpo-33670: Expose Sphinx errorlevel (GH-7156)

vstinner committed:
bpo-33674: asyncio: Fix SSLProtocol race (GH-7175)

Fix a race condition in SSLProtocol.connection_made() of
asyncio.sslproto: start immediately the handshake instead of using
call_soon(). Previously, data_received() could be called before the
handshake started, causing the handshake to hang or fail.

zooba committed:
bpo-32878: Adds documentation for st_ino on Windows (GH-5764)

Commit:
bpo-33674: Pause the transport as early as possible (#7192)

Commit:
bpo-33654: Support protocol type switching in SSLTransport.set_protocol() (#7194)

serhiy-storchaka committed:
bpo-32911: Revert bpo-29463. (GH-7121) (GH-7197)

Remove the docstring attribute of AST types and restore docstring
expression as a first stmt in their body.

Co-authored-by: INADA Naoki <methane@users.noreply.github.com>

twisteroidambassador committed:
Use module import instead of individual imports.

twisteroidambassador committed:
Use module import instead of individual imports.

twisteroidambassador committed:
Change TODO.

twisteroidambassador committed:
Change TODO.

twisteroidambassador committed:
Rename helpers.py to staggered.py.

twisteroidambassador committed:
Rename helpers.py to staggered.py.

twisteroidambassador committed:
Add create_connection()'s new arguments in documentation.

twisteroidambassador committed:
Add create_connection()'s new arguments in documentation.

twisteroidambassador committed:
Add blurb.

twisteroidambassador committed:
Add blurb.

vstinner committed:
bpo-33353: test_asyncio use set_write_buffer_limits() (GH-7200)

Use transport.set_write_buffer_limits() in sendfile tests of
test_asyncio to make sure that the protocol is paused after sending
4 KiB. Previously,
test_sendfile_fallback_close_peer_in_the_middle_of_receiving() failed
on FreeBSD if the DATA was smaller than the default limit of 64 KiB.

corona10 committed:
bpo-33197: Update a error message of invalid inspect.Parameters. (GH-6636)

serhiy-storchaka committed:
bpo-33622: Add checks for exceptions leaks in the garbage collector. (GH-7126)

* Failure in adding to gc.garbage is no longer fatal.
* An exception in tp_clear() no longer lead to crash (though tp_clear() should not leave exceptions).

Commit:
Attempt to fix test_stdin_broken_pipe on Travis (#7210)

daxlab committed:
bpo-30935: update get_event_loop docs (GH-2731)

Commit:
bpo-22087: Fix Policy.get_event_loop() to detect fork (GH-7208)

Original patch by Dan O'Reilly.

geryogam committed:
bpo-31639: Change ThreadedHTTPServer to ThreadingHTTPServer class name (GH-7195)

Commit:
bpo-32684: Fix gather to propagate cancel of itself with return_exceptions (GH-7209)

elprans committed:
bpo-32751: Wait for task cancellation in asyncio.wait_for() (GH-7216)

Currently, asyncio.wait_for(fut), upon reaching the timeout deadline,
cancels the future and returns immediately.  This is problematic for
when *fut* is a Task, because it will be left running for an arbitrary
amount of time.  This behavior is iself surprising and may lead to
related bugs such as the one described in bpo-33638:

    condition = asyncio.Condition()
    async with condition:
        await asyncio.wait_for(condition.wait(), timeout=0.5)

Currently, instead of raising a TimeoutError, the above code will fail
with `RuntimeError: cannot wait on un-acquired lock`, because
`__aexit__` is reached _before_ `condition.wait()` finishes its
cancellation and re-acquires the condition lock.

To resolve this, make `wait_for` await for the task cancellation.
The tradeoff here is that the `timeout` promise may be broken if the
task decides to handle its cancellation in a slow way.  This represents
a behavior change and should probably not be back-patched to 3.6 and
earlier.

elprans committed:
bpo-23859: Document that asyncio.wait() does not cancel its futures (#7217)

Unlike `asyncio.wait_for()`, `asyncio.wait()` does not cancel the passed
futures when a timeout accurs.

JulienPalard committed:
Fix whatsnew entry about ThreadedHTTPServer. (GH-7220)

Commit:
bpo-32684: Fix nits in tests (GH-7225)

elprans committed:
Fix typo introduced in GH-7217 (#7230)

Commit:
Revert "bpo-22087: Fix Policy.get_event_loop() to detect fork (GH-7208)" (#7232)

This reverts commit 5d97b7bcc19496617bf8c448d2f149cc28c73bc7.

Commit:
A better fix for asyncio test_stdin_broken_pipe (#7221)

Commit:
Revert "A better fix for asyncio test_stdin_broken_pipe (GH-7221)" (GH-7235)

This reverts commit ad74d50517c069359578014bb32e660828d68332.

Turns out it's not a good fix -- Travis has just crashed on this test.

twisteroidambassador committed:
Merge remote-tracking branch 'origin/happy-eyeballs' into happy-eyeballs
